File: 238. Product of Array Except Self/main.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        size = len(nums)
        prefix = [1] * size
        suffix = [1] * size

        for i in range(1, size):
            prefix[i] = prefix[i - 1] * nums[i - 1]

        for i in range(size - 2, -1, -1):
            suffix[i] = suffix[i + 1] * nums[i + 1]

        answer = [1] * size
        for i in range(size):
            answer[i] = prefix[i] * suffix[i]

        return answer


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        size = len(nums)
        answer = [1] * size

        prefix = 1
        for i in range(size):
            answer[i] = prefix
            prefix *= nums[i]

        suffix = 1
        for i in range(size - 1, -1, -1):
            answer[i] *= suffix
            suffix *= nums[i]

        return answer

File: 238. Product of Array Except Self/main.rs
struct Solution;

impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let size = nums.len();
        let mut prefix = vec![1; size];
        let mut suffix = vec![1; size];

        for i in 1..size {
            prefix[i] = prefix[i - 1] * nums[i - 1];
        }

        for i in (0..size - 1).rev() {
            suffix[i] = suffix[i + 1] * nums[i + 1];
        }

        let mut answer = vec![1; size];
        for i in 0..size {
            answer[i] = prefix[i] * suffix[i];
        }

        answer
    }
}

impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let size = nums.len();
        let mut prefix = vec![1; size];
        let mut suffix = vec![1; size];

        for i in 1..size {
            prefix[i] = prefix[i - 1] * nums[i - 1];
        }

        for i in (0..size - 1).rev() {
            suffix[i] = suffix[i + 1] * nums[i + 1];
        }

        (0..size).map(|i| prefix[i] * suffix[i]).collect()
    }
}

impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let size = nums.len();
        let mut answer = vec![1; size];
        let mut prefix = 1;

        for i in 0..size {
            answer[i] = prefix;
            prefix *= nums[i];
        }

        let mut suffix = 1;
        for i in (0..size).rev() {
            answer[i] *= suffix;
            suffix *= nums[i];
        }

        answer
    }
}
